File: tools/webcam/webcam.cc
#include <iostream>
#include <chrono>
#include <thread>
#include <cv.h>
#include <highgui.h>

#include "cereal/gen/cpp/log.capnp.h"
#include <capnp/message.h>
#include <capnp/serialize.h>
#include <kj/array.h>
#include <kj/io.h>

#include "messaging.hpp"

#define FRAME_WIDTH  1164
#define FRAME_HEIGHT 874

Context* context = Context::create();

void* health_function(void*) {
  PubSocket* sock = PubSocket::create(context, "health");

  std::chrono::seconds sec(1);
  std::chrono::steady_clock::time_point next_time = std::chrono::steady_clock::now() + sec;
  while (1) {
    // Build capnp message
    capnp::MallocMessageBuilder message;

    cereal::Event::Builder event_builder = message.initRoot<cereal::Event>();
    event_builder.setLogMonoTime(0); // TODO

    cereal::HealthData::Builder health_builder = event_builder.initHealth();
    health_builder.setIgnitionLine(true);
    health_builder.setHwType(cereal::HealthData::HwType::WHITE_PANDA);
    health_builder.setControlsAllowed(true);

    // Serialize message
    kj::VectorOutputStream stream;
    capnp::writeMessage(stream, message);
    kj::ArrayPtr<unsigned char> stream_array = stream.getArray();

    // Send serialized message through cereal messaging 
    sock->send(reinterpret_cast<char*>(stream_array.begin()), stream_array.size());
    
    std::this_thread::sleep_until(next_time);
    next_time += sec;
  }

  return 0;
}

int frame_function() {
  cv::VideoCapture cap(0);
  PubSocket* sock = PubSocket::create(context, "frame");

  if (!cap.isOpened()) {
    std::cout << "Error opening VideoCapture" << std::endl;
    return -1;
  }

  while (1) {
    cv::Mat frame_mat;

    cap >> frame_mat;
    if (frame_mat.empty())
      break;

    cv::resize(frame_mat, frame_mat, cv::Size(FRAME_WIDTH, FRAME_HEIGHT));

    // Get image data size
    int frame_size = frame_mat.total() * frame_mat.elemSize();

    // Get raw image data
    unsigned char* frame_buffer = new unsigned char[frame_size];
    memcpy(frame_buffer, frame_mat.data, frame_size * sizeof(char));

    // Build capnp message
    capnp::MallocMessageBuilder message;

    cereal::Event::Builder event_builder = message.initRoot<cereal::Event>();
    event_builder.setLogMonoTime(0); // TODO

    cereal::FrameData::Builder frame_builder = event_builder.initFrame();
    frame_builder.setImage(kj::arrayPtr(frame_buffer, frame_size));

    // Serialize message
    kj::VectorOutputStream stream;
    capnp::writeMessage(stream, message);
    kj::ArrayPtr<unsigned char> stream_array = stream.getArray();

    // Send serialized message through cereal messaging 
    sock->send(reinterpret_cast<char*>(stream_array.begin()), stream_array.size());

    // Clean up memory
    delete[] frame_buffer;
  }

  cap.release();
  cv::destroyAllWindows();
  return 0;
}

int main() {
  pthread_t thread_handle;
  pthread_create(&thread_handle, 0, health_function, 0);
  pthread_detach(thread_handle);

  frame_function();

  return 0;
}

File: tools/webcam/webcam.py
#!/usr/bin/env python3
# This file does the same thing as webcam.cc, but in python...
# It is now obsolete but it is useful for prototyping.
from common.realtime import Ratekeeper
import cereal.messaging as messaging
import numpy as np
import threading
import cv2

FRAME_WIDTH, FRAME_HEIGHT = 1164, 874

def health_function():
  pm = messaging.PubMaster(['health'])
  rk = Ratekeeper(1.0)
  while (True):
    dat = messaging.new_message()
    dat.init('health')
    dat.valid = True
    dat.health = {
      'ignitionLine': True,
      'hwType': "whitePanda",
      'controlsAllowed': True
    }
    pm.send('health', dat)
    rk.keep_time()

def frame_function():
  cap = cv2.VideoCapture(0)
  pm = messaging.PubMaster(['frame'])

  while (True):
    ret, img = cap.read()
    
    if ret:
      img = cv2.resize(img, (FRAME_WIDTH, FRAME_HEIGHT))

      dat = messaging.new_message()
      dat.init('frame')
      dat.frame = {
        #"frameId": frame_id,
        "image": img.tostring(),
      }
      pm.send('frame', dat)

if __name__=="__main__":
  threading.Thread(target=health_function).start()
  frame_function()

  cap.release()
  cv2.destroyAllWindows()
